eventsources/sources/bitbucket/start.go

```diff
@@ -0,0 +1,344 @@
+
+	logger.Info("listing existing webhooks...")
+	hooks, err := router.listWebhooks()
+	if err != nil {
+		logger.Errorw("failed to list webhooks", zap.Error(err))
+		return errors.Wrap(err, "failed to list webhooks")
+	}
+
+	logger.Info("checking if webhook already exists...")
+	existingHookSubscription, isFound := router.findWebhook(hooks, formattedWebhookURL)
+	if isFound && router.shouldUpdateWebhook(existingHookSubscription) {
+		logger.Info("webhook already exists but requires an update...")
+		if _, err = router.updateWebhook(existingHookSubscription); err != nil {
+			logger.Errorw("failed to update existing webhook", zap.Error(err))
+			return errors.Wrap(err, "failed to update existing webhook")
+		}
+
+		logger.Info("successfully updated the existing webhook")
+		return nil
+	}
+
+	logger.Info("webhook doesn't exist yet, creating a new webhook...")
+	newWebhook, err := router.createWebhook(formattedWebhookURL)
+	if err != nil {
+		logger.Errorw("failed to create new webhook", zap.Error(err))
+		return errors.Wrap(err, "failed to create new webhook")
+	}
+
+	router.hookID = newWebhook.Uuid
+
+	logger.Info("successfully created a new webhook")
+	return nil
+}
+
+// createWebhook creates a new webhook
+func (router *Router) createWebhook(formattedWebhookURL string) (*bitbucketv2.Webhook, error) {
+	es := router.bitbucketEventSource
+	opt := &bitbucketv2.WebhooksOptions{
+		Owner:       es.Owner,
+		RepoSlug:    es.RepositorySlug,
+		Url:         formattedWebhookURL,
+		Description: "webhook managed by Argo-Events",
+		Active:      true,
+		Events:      es.Events,
+	}
+
+	return router.client.Repositories.Webhooks.Create(opt)
+}
+
+// updateWebhook updates an existing webhook
+func (router *Router) updateWebhook(existingHookSubscription *WebhookSubscription) (*bitbucketv2.Webhook, error) {
+	es := router.bitbucketEventSource
+	opt := &bitbucketv2.WebhooksOptions{
+		Owner:       es.Owner,
+		RepoSlug:    es.RepositorySlug,
+		Uuid:        existingHookSubscription.Uuid,
+		Description: existingHookSubscription.Description,
+		Url:         existingHookSubscription.Url,
+		Active:      existingHookSubscription.Active,
+		Events:      es.Events,
+	}
+
+	return router.client.Repositories.Webhooks.Update(opt)
+}
+
+// deleteWebhook deletes an existing webhook
+func (router *Router) deleteWebhook(hookID string) error {
+	es := router.bitbucketEventSource
+	_, err := router.client.Repositories.Webhooks.Delete(&bitbucketv2.WebhooksOptions{
+		Owner:    es.Owner,
+		RepoSlug: es.RepositorySlug,
+		Uuid:     hookID,
+	})
+
+	return err
+}
+
+// listWebhooks gets a list of all existing webhooks in target repository
+func (router *Router) listWebhooks() ([]WebhookSubscription, error) {
+	es := router.bitbucketEventSource
+	hooksResponse, err := router.client.Repositories.Webhooks.Gets(&bitbucketv2.WebhooksOptions{
+		Owner:    es.Owner,
+		RepoSlug: es.RepositorySlug,
+		Active:   true,
```

Comment:
Should we not list all webhooks, active or not? If someone manually deactivates the webhook registered on the BitBucket and the event source has `deleteOnFinish` set to `false`, then the code will try to register it as a fresh webhook when eventsource pod restarts. Instead, we just need to update the existing one.

Reply:
Yes, totally right. Actually the `Gets` method only uses `Owner` and `RepoSlug` for the query, I just wanted to test something and forgot to remove it.
Fixing now.